src/Compilers/CSharp/Portable/Symbols/PublicModel/NamedTypeSymbol.cs

```diff
@@ -0,0 +1,202 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+
+namespace Microsoft.CodeAnalysis.CSharp.Symbols.PublicModel
+{
+    internal abstract class NamedTypeSymbol : TypeSymbol, INamedTypeSymbol
+    {
+        private ImmutableArray<ITypeSymbol> _lazyTypeArguments;
+
+        public NamedTypeSymbol(CodeAnalysis.NullableAnnotation nullableAnnotation = CodeAnalysis.NullableAnnotation.None)
+            : base(nullableAnnotation)
+        {
+        }
+
+        internal abstract Symbols.NamedTypeSymbol UnderlyingNamedTypeSymbol { get; }
+
+        int INamedTypeSymbol.Arity
+        {
+            get
+            {
+                return UnderlyingNamedTypeSymbol.Arity;
+            }
+        }
+
+        ImmutableArray<IMethodSymbol> INamedTypeSymbol.InstanceConstructors
+        {
+            get
+            {
+                return UnderlyingNamedTypeSymbol.InstanceConstructors.GetPublicSymbols();
+            }
+        }
+
+        ImmutableArray<IMethodSymbol> INamedTypeSymbol.StaticConstructors
+        {
+            get
+            {
+                return UnderlyingNamedTypeSymbol.StaticConstructors.GetPublicSymbols();
+            }
+        }
+
+        ImmutableArray<IMethodSymbol> INamedTypeSymbol.Constructors
+        {
+            get
+            {
+                return UnderlyingNamedTypeSymbol.Constructors.GetPublicSymbols();
+            }
+        }
+
+        IEnumerable<string> INamedTypeSymbol.MemberNames
+        {
+            get
+            {
+                return UnderlyingNamedTypeSymbol.MemberNames;
+            }
+        }
+
+        ImmutableArray<ITypeParameterSymbol> INamedTypeSymbol.TypeParameters
+        {
+            get
+            {
+                return UnderlyingNamedTypeSymbol.TypeParameters.GetPublicSymbols();
+            }
+        }
+
+        ImmutableArray<ITypeSymbol> INamedTypeSymbol.TypeArguments
+        {
+            get
+            {
+                if (_lazyTypeArguments.IsDefault)
+                {
+
+                    ImmutableInterlocked.InterlockedCompareExchange(ref _lazyTypeArguments, UnderlyingNamedTypeSymbol.TypeArgumentsWithAnnotationsNoUseSiteDiagnostics.SelectAsArray(t => t.GetITypeSymbol()), default);
```

Comment:
>t => t.GetITypeSymbol() [](start = 181, length = 23)

We use equivalent lambdas in a few places, it might be worth caching an internal static readonly copy on the public `TypeSymbol` and just using that everywhere.

Reply:
> We use equivalent lambdas in a few places, it might be worth caching an internal static readonly copy on the public TypeSymbol and just using that everywhere.

I will extract a helper for translating array of TypeWithAnnotations.

---
In reply to: [338826951](https://github.com/dotnet/roslyn/pull/39498#discussion_r338826951) [](ancestors = 338826951)